File: app/src/main/java/com/aaron/yespdf/filepicker/ViewAllPresenter.java
package com.aaron.yespdf.filepicker;

import com.blankj.utilcode.util.SDCardUtils;
import com.blankj.utilcode.util.StringUtils;

import java.io.File;
import java.util.Arrays;
import java.util.List;

/**
 * @author Aaron aaronzzxup@gmail.com
 */
class ViewAllPresenter extends IViewAllContract.P {

    private String curPath = ROOT_PATH;
    private boolean isFirstIn = true;

    ViewAllPresenter(IViewAllContract.V v) {
        super(v);
    }

    @Override
    IViewAllContract.M model() {
        return new ViewAllModel();
    }

    @Override
    void detachV() {
        model.saveLastPath(curPath);
        view = null;
        model = null;
    }

    @Override
    boolean canFinish() {
        return curPath.equals(ROOT_PATH);
    }

    @Override
    void goBack() {
        String prePath = curPath.substring(0, curPath.lastIndexOf("/"));
        listFile(prePath);
    }

    @Override
    void listStorage() {
        if (isFirstIn) {
            isFirstIn = false;
            String lastPath = model.queryLastPath();
            if (StringUtils.isEmpty(lastPath) || ROOT_PATH.equals(lastPath)) {
                listStorage();
            } else {
                listFile(lastPath);
            }
            return;
        }
        model.listStorage(new IViewAllContract.FileCallback<List<SDCardUtils.SDCardInfo>>() {
            @Override
            public void onResult(List<SDCardUtils.SDCardInfo> result) {
                fileList.clear();
                for (SDCardUtils.SDCardInfo info : result) {
                    if (info.getState().equals("mounted")) {
                        fileList.add(new File(info.getPath()));
                    }
                }
                view.onShowFileList(fileList);
            }
        });
    }

    @Override
    void listFile(final String path) {
        curPath = path;
        if (path.equals(ROOT_PATH)) {
            listStorage();
            return;
        }
        model.listFile(path, new IViewAllContract.FileCallback<List<File>>() {
            @Override
            public void onResult(List<File> result) {
                if (path.length() > 18) {
                    String path1 = path.substring(18); // 去除 /storage/emulated/
                    List<String> paths = Arrays.asList(path1.split("/"));
                    view.onShowPath(paths);
                }
                view.onShowFileList(result);
            }
        });
    }
}

File: app/src/main/java/com/aaron/yespdf/filepicker/ViewAllPresenter.kt
package com.aaron.yespdf.filepicker

import com.blankj.utilcode.util.StringUtils
import java.io.File
import java.util.*

/**
 * @author Aaron aaronzzxup@gmail.com
 */
class ViewAllPresenter(view: IViewAllView) : IViewAllPresenter(view) {

    private var curPath: String? = ROOT_PATH
    private var isFirstIn = true

    override val model: IViewAllModel = ViewAllModel()

    override fun detach() {
        model.saveLastPath(curPath)
    }

    override fun canFinish(): Boolean {
        return curPath == ROOT_PATH
    }

    override fun goBack() {
        val prePath = curPath?.substring(0, curPath?.lastIndexOf("/") ?: 0)
        listFile(prePath)
    }

    override fun listStorage() {
        if (isFirstIn) {
            isFirstIn = false
            val lastPath = model.queryLastPath()
            if (StringUtils.isEmpty(lastPath) || ROOT_PATH == lastPath) {
                listStorage()
            } else {
                listFile(lastPath)
            }
            return
        }
        model.listStorage { result ->
            fileList.clear()
            if (result != null) {
                for (info in result) {
                    if (info.state == "mounted") {
                        fileList.add(File(info.path))
                    }
                }
            }
            view.onShowFileList(fileList)
        }
    }

    override fun listFile(path: String?) {
        curPath = path
        if (path == ROOT_PATH) {
            listStorage()
            return
        }
        model.listFile(path) { result ->
            path?.run {
                if (length > 18) {
                    val path1 = substring(18) // 去除 /storage/emulated/
                    val paths = listOf(*path1.split("/").toTypedArray())
                    view.onShowPath(paths)
                }
            }
            view.onShowFileList(result)
        }
    }
}